File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/structdefutil/LHArrayType.java
package io.littlehorse.sdk.wfsdk.internal.structdefutil;

import io.littlehorse.sdk.common.adapter.LHTypeAdapterRegistry;
import io.littlehorse.sdk.common.proto.InlineArrayDef;
import io.littlehorse.sdk.common.proto.TypeDefinition;
import io.littlehorse.sdk.common.proto.TypeDefinition.DefinedTypeCase;
import java.util.Map;

/**
 * Represents a native LittleHorse array type.
 */
public final class LHArrayType extends LHClassType {

    private final LHClassType componentType;

    public LHArrayType(Class<?> clazz, LHTypeAdapterRegistry typeAdapterRegistry) {
        this(clazz, typeAdapterRegistry, Map.of());
    }

    public LHArrayType(
            Class<?> clazz, LHTypeAdapterRegistry typeAdapterRegistry, Map<String, String> placeholderValues) {
        super(clazz, typeAdapterRegistry, placeholderValues);

        if (!clazz.isArray()) {
            throw new IllegalArgumentException(
                    "LHArrayType can only be created from array classes. Provided class: " + clazz.getName()
                            + ". Please add brackets to your class to declare an array type, e.g. MyType[] instead of MyType.");
        }
        if (byte[].class.equals(clazz)) {
            throw new IllegalArgumentException(
                    "byte[] is not supported as an array type in LittleHorse. This type has special handling as a primitive type. If you want an array of multiple bytes items, consider using `byte[][]`.");
        }

        Class<?> componentClass = clazz.getComponentType();

        if (componentClass.isArray()) {
            this.componentType = new LHArrayType(componentClass, typeAdapterRegistry, this.placeholderValues);
        } else {
            this.componentType = LHClassType.fromJavaClass(componentClass, typeAdapterRegistry, this.placeholderValues);
        }

        try {
            LHTypeConstraintValidator.ensureNoJsonPrimitiveTypes(this.componentType.getTypeDefinition());
        } catch (ForbiddenJsonTypeException ex) {
            throw new IllegalArgumentException(
                    String.format(
                            "InlineArrayDef element type %s for Java array %s: %s",
                            componentClass.getCanonicalName(), clazz.getCanonicalName(), ex.getMessage()),
                    ex);
        }
    }

    @Override
    public DefinedTypeCase getDefinedTypeCase() {
        return DefinedTypeCase.INLINE_ARRAY_DEF;
    }

    @Override
    public TypeDefinition getTypeDefinition() {
        return TypeDefinition.newBuilder()
                .setInlineArrayDef(InlineArrayDef.newBuilder().setArrayType(componentType.getTypeDefinition()))
                .build();
    }
}

File: sdk-dotnet/LittleHorse.Sdk/Helper/LHArrayType.cs
using LittleHorse.Sdk.Common.Proto;
using LittleHorse.Sdk.Exceptions;

namespace LittleHorse.Sdk.Helper
{
    /// <summary>
    /// Represents a native LittleHorse array type.
    /// </summary>
    public class LHArrayType : LHClassType
    {
        private readonly LHClassType _componentType;

        /// <summary>
        /// Creates an LH array wrapper from a .NET array type.
        /// </summary>
        /// <param name="type">A .NET array type (for example, typeof(string[])).</param>
        public LHArrayType(Type type) : base(type)
        {
            if (!type.IsArray)
            {
                throw new ArgumentException($"LHArrayType can only be created from array types; got {type.Name}.");
            }

            if (type == typeof(byte[]))
            {
                throw new ArgumentException("byte[] is BYTES in LittleHorse and cannot be used as a native LH Array type.");
            }

            Type? componentClass = type.GetElementType();
            if (componentClass == null)
            {
                throw new ArgumentException($"Unable to resolve array component type for {type.Name}.");
            }

            _componentType = componentClass.IsArray
                ? new LHArrayType(componentClass)
                : LHClassType.FromType(componentClass);

            try
            {
                LHTypeConstraintValidator.EnsureNoJsonPrimitiveTypes(_componentType.GetTypeDefinition());
            }
            catch (ForbiddenJsonTypeException ex)
            {
                throw new ArgumentException(
                    $"Native LH Array component type cannot contain {ex.ForbiddenType}. Use strongly typed Struct/Array definitions.",
                    ex);
            }
        }

        /// <inheritdoc />
        public override TypeDefinition.DefinedTypeOneofCase GetDefinedTypeCase()
        {
            return TypeDefinition.DefinedTypeOneofCase.InlineArrayDef;
        }

        /// <inheritdoc />
        public override TypeDefinition GetTypeDefinition()
        {
            return new TypeDefinition
            {
                InlineArrayDef = new InlineArrayDef
                {
                    ArrayType = _componentType.GetTypeDefinition()
                }
            };
        }
    }
}
